File: lol_gen_alg_frontend/src/gen_alg_logic/genetic_algorithm.js
import TeamComp from './team_comp.js';
import cloneDeep from 'lodash/cloneDeep';


export default class GeneticAlgorithm {
  constructor(
    population, crossover_rate,
    mutation_rate, enemy_team_comp, champ_info, meta=true
  ){
    this.champ_info = champ_info;
    this.crossover_rate = crossover_rate;
    this.enemy_team_comp = enemy_team_comp;
    this.best_individuals = [];
    this.iteration_data = [];
    this.average_fitness = 0;

    this.population = [];
    for (let i = 0; i < population; i ++) {
      this.population.push(new TeamComp(champ_info, null, mutation_rate, meta));
    }
  }

  crossover() {
    /*
    Apply crossover as per the crossover ratio.
    Crossover will be to swap to champions at a random
    index for two individuals in the population
    */
    for (let i = 0; i < this.population.length; i ++) {
      // Check if we perform crossover
      if (Math.random() <= this.crossover_rate){
        let temp_index = getRandomInt(this.population);
        while(temp_index === i){
          temp_index = getRandomInt(this.population);
        }

        // Temp assignment for givien individual
        let curr_individual = this.population[i];
        let temp_individual = this.population[temp_index];

        // Try crossover
        let valid_crossover = false;
        let tries = 0;

        // Only attempting for the size of the curr_individual champions (should be 5)
        while(!valid_crossover && tries < curr_individual) {
          let champ_index = getRandomInt(5);

          // Check that this swap can be complete for both teams
          let is_on_curr = curr_individual.is_on_team(temp_individual.champions[champ_index].id);
          let is_on_temp = temp_individual.is_on_team(curr_individual.champions[champ_index].id);

          if (!(is_on_curr || is_on_temp)){
            // Then we can swap
            let temp = curr_individual.champions[champ_index];
            curr_individual.champions[champ_index] = temp_individual.champions[champ_index];
            temp_individual.champions[champ_index] = temp;
            valid_crossover = true;
          }
          else {
            tries++;
          }
        }
      }
    }
  }

  mutate(){
    for (let i = 0; i < this.population.length; i++){
      this.population[i].mutate(this.champ_info);
    }
  }

  selection_tournament(){
    // Calculate fitness first
    for (let i = 0; i < this.population.length; i++){
      this.population[i].calculate_fitness(this.enemy_team_comp.champions);
    }

    let new_generation = [];
    // 1v1 Selection after calculating the fitness
    for (let i = 0; i < this.population.length; i++){
      let temp_index = getRandomInt(this.population.length);

      while (temp_index === i) {
        temp_index = getRandomInt(this.population.length);
      }

      if (this.population[i].fitness >= this.population[temp_index].fitness){
        new_generation.push(cloneDeep(this.population[i]));
      }
      else {
        new_generation.push(cloneDeep(this.population[temp_index]));
      }
    }

    this.population = new_generation;
  }

  calculate_best(n_generation){
    // Calculate average fitness
    var avg_fitness = 0
    for (let i = 0; i < this.population.length; i++){
      avg_fitness = avg_fitness + this.population[i].fitness;
    }
    this.average_fitness = avg_fitness/this.population.length;
    this.iteration_data.push({
      generations: n_generation,
      fitness: this.average_fitness
    });
    // Keep list of 10 best individuals.
    this.best_individuals = this.best_individuals.concat(this.population);
    this.best_individuals = this.best_individuals.sort((a,b) => (a.fitness > b.fitness) ? -1 : 1);
    this.best_individuals = this.best_individuals.slice(0,10);
  }

  iterate_population(n_generation){
    this.crossover();
    this.mutate();
    this.selection_tournament();
    this.calculate_best(n_generation);
  }

}

function getRandomInt(max) {
 	return Math.floor(Math.random() * Math.floor(max));
}

File: lol_gen_alg_python/genetic_algorithm/genetic_algorithm.py
from genetic_algorithm.team_comp import TeamComp
import random
import logging
import copy

logger = logging.getLogger()


class GeneticAlgorithm:
    def __init__(self, population, crossover_rate, mutation_rate,
                 base_team_composition: TeamComp, champion_information: dict, meta=True):
        """
        Genetic Algorithm to calculate best team composition to counter given team
        :param population: Size of Population
        :param crossover_rate: Crossover Rate
        :param mutation_rate: Mutation Rate
        :param base_team_composition: Team Composition we are optimizing against
        :param champion_information: Available champion_information
        :param meta: make a meta team comp
        """
        logger.info(f"Population: {population} Crossover: {crossover_rate} Mutation Rate: {mutation_rate}")
        self.champion_information = champion_information

        self.crossover_rate = crossover_rate
        self.base_team_composition = base_team_composition
        self.best_individuals = []
        self.average_fitness = []

        self.population = []
        for i in range(population):
            self.population.append(TeamComp(champion_information, champions=None, meta=meta))

    def crossover(self):
        """
        Apply crossover as per the crossover ratio.
        Crossover will be to swap to champions at a random
        index for two individuals in the population
        """
        for curr_index in range(len(self.population)):
            # Check if we perform crossover
            if random.uniform(0, 1) <= self.crossover_rate:
                # Make sure that we don't select current individual
                temp_index = random.randint(0, len(self.population)-1)
                while temp_index == curr_index:
                    temp_index = random.randint(0, len(self.population)-1)

                # Temp assignment for given individual
                curr_individual = self.population[curr_index]
                temp_individual = self.population[temp_index]

                # Try to crossover
                valid_crossover = False
                tries = 0

                # Only attempting for the size of the curr_individual champions (should be 5)
                while not valid_crossover and tries < len(curr_individual.champions):
                    champ_index = random.randint(0, 4)
                    # Check that this swap can be complete for both teams
                    is_on_curr = curr_individual.is_on_team(temp_individual.champions[champ_index].id)
                    is_on_temp = temp_individual.is_on_team(curr_individual.champions[champ_index].id)

                    if not (is_on_curr or is_on_temp):
                        # Then we can swap
                        temp = curr_individual.champions[champ_index]
                        curr_individual.champions[champ_index] = temp_individual.champions[champ_index]
                        temp_individual.champions[champ_index] = temp
                        logger.debug(f"Swapped {curr_individual.champions[champ_index].name} to {curr_index} "
                                     f"and {temp_individual.champions[champ_index].name} to {temp_index}")
                        valid_crossover = True
                    else:
                        tries += 1

    def mutate(self):
        """
        We mutate each individual in the population
        :return:
        """
        for i in range(len(self.population)):
            self.population[i].mutate(
                available_champions=self.champion_information,
                individual=i)

    def selection_tournament(self):
        """
        Selection Tournament with Replacement
        :return:
        """
        for i in range(len(self.population)):
            self.population[i].calculate_fitness(self.base_team_composition.champions)
        fitness_vals = ",".join([str(i.fitness) for i in self.population])
        logger.info(f"New fitness values: {fitness_vals}")

        new_generation = []
        # Currently only 1v1 selection
        for i in range(len(self.population)):
            # Make sure that we don't select current individual
            individual1 = random.randint(0, len(self.population)-1)
            while individual1 == i:
                individual1 = random.randint(0, len(self.population)-1)

            if self.population[i].fitness >= self.population[individual1].fitness:
                new_generation.append(copy.deepcopy(self.population[i]))

            else:
                new_generation.append(copy.deepcopy(self.population[individual1]))

        self.population = new_generation

    def calculate_best(self):
        avg_fitness = 0
        for i in self.population:
            avg_fitness += i.fitness
        self.average_fitness.append(avg_fitness/len(self.population))
        logger.info(f"Average Fitness: {self.average_fitness[-1]}")

        self.best_individuals += self.population
        self.best_individuals.sort(key=lambda x: x.fitness, reverse=True)
        self.best_individuals = self.best_individuals[:10]
        fitness_vals = ",".join([str(i.fitness) for i in self.best_individuals])
        logger.info(f"Best Fitness: {fitness_vals}")


    def iterate_population(self):
        self.crossover()
        self.mutate()
        self.selection_tournament()
        self.calculate_best()
